### ISSUE TITLE
import/export full session

### ISSUE THREAD

**Issue by tgirod** [fariapp/fari#148]
## 🚩 Is your feature request related to a problem? Please describe

Right now, saving a complete session can be a hassle. You have to save each character and each scene separately. If you use fari to run multiple games, it can be frustrating.

## ✅ Describe the solution you'd like

Being able to import/export a full session (scenes + characters) would be of great help to manage multiple games, or just to save your game if you don't really trust your browser's local storage. The import dialog would need a warning: "importing a session will overwrite the existing one. Are you sure?"

## ✏️ Describe alternatives you've considered

saving each scene and character by hand.

**Comment by tgirod** [comment]
Here is some feedback from my session zero. Things didn't go too well. My 3 players joined the game with the character sheet I sent them. We edited them together. We had a few times when a modification took some time to appear everywhere, but appart from that it went ok.

The problem arrived when we finished. I asked my players to save their sheets because I couldn't do it for them. Of my 3 players, two somehow saved unmodified sheets, and lost all the work we did.

I think having the GM as the "single source of truth" for all the session's data, and giving him/her the ability to save and restore everything easily is really necessary.

**Comment by RPDeshaies** [comment]
When did this happen exactly ?

**Comment by RPDeshaies** [comment]
The reason I'm asking is that there was a pretty big sync issue in the code that I fixed in the latest release that I deployed this morning.

Basically, I was not comparing "lastUpdatedTimes" propertly between character sheets and then it caused some issues. 

Things should be way more stable now

**Comment by tgirod** [comment]
It happened last monday (9th of november), on the previous release.

**Comment by RPDeshaies** [comment]
Alright very good to know. Things should be way more stable now and I'm very sorry you experienced this as this is a pretty big issue.

---

About the main issue #148, it's not something that would be feasible in the short term. 
Let me try to add a bit more details as to why.

Like in real life, players are expected to manage their character sheets themselves.

Fari manages Scenes and Characters but those two entities are saved separately and even when Characters are added to the Scene, they are added to an ephemeral property `scene.players.[index].character` that is not part of what I call in the code "A Savable Scene Property".

To implement what you are asking, I would need to let the GM export the full scene, but when imported again it would need to match the character sheets that were exported to the character sheet currently in session.

It's would be a bit of a hassle and could cause some sync issues.

Though what I could do though would be to split "player character sheets" from the ones the GM can add from the "add offline character sheet" button and make that entity part of the "savable scene". That would result in GMs managing character sheets themselves but every time a new scene is created or even reinitialized, the character sheets would need to be added manually to the scene again.

Mostly thinking out loud here ☝️ but it's not an easy problem to solve I hope this explanation adds a bit more detail as to why.

Anyhow, it's still something I'm going to keep in mind and if I come up with a nice idea, I'll update this issue.

**Comment by tgirod** [comment]
Thanks for the explanation.

> Like in real life, players are expected to manage their character sheets themselves.

Not on many tables I've participated to - as a GM I usually centralize my game's materials, including character sheets. So they are not lost or forgotten at home, and I have them as a reference during game prep if necessary. I'm not saying it's better, it's just to give some perspective on the subject.

But I get this is not the way you designed your application, so it is not an easy thing to bolt on. I guess it would be necessary to redefine a root object `Game`, which contains scenes and characters. I guess it could turn into quite a refactoring.

**Comment by RPDeshaies** [comment]
>  I guess it could turn into quite a refactoring.

Indeed, especially considering that now that people have scenes and characters already saved in their local storage, we would need to do a data migration on page load to changes the data structure.

The migration mechanisms are already in place in the code but that doesn't eliminate the time it takes to write and test those migrations.

I really hope that the fix I deployed yesterday will fix the issue you describe earlier as I believe that the current structure in fine, as long as it's stable and reliable obviously 😅 

I recently incorporated a new testing framework so in the future I'll try to add more tests for those kinds of issues.
